cql3/expr/expression.cc

```diff
@@ -876,16 +939,15 @@ expression replace_column_def(const expression& expr, const column_definition* n
                 return expression(conjunction{std::vector(applied.begin(), applied.end())});
             },
             [&] (const binary_operator& oper) {
-                return std::visit(overloaded_functor{
-                        [&] (const column_value& col) {
-                            return expression(binary_operator{column_value{new_cdef}, oper.op, oper.rhs});
-                        },
-                        [&] (const column_value_tuple& tuple) -> expression {
-                            throw std::logic_error(format("replace_column_def invalid LHS: {}", to_string(oper)));
-                        },
-                        [&] (const token&) { return expr; },
-                    }, oper.lhs);
+                return expression(binary_operator(replace_column_def(*oper.lhs, new_cdef), oper.op, oper.rhs, oper.order));
```

Comment:
I did not intend to fix a bug here, and such bug fixes should not be hidden in large patches. I'll send a separate fix for this, and then rebase my patch on top.

Reply:
Converted to draft pending this bugfix.

Reply:
This is more involved. This function is used when converting a restriction on some column in the base table to a restriction on the materialized view that backs the index.

If the restriction is equality, then `order` doesn't affect anything.

If the restriction is ordering, then there are two cases:

1. it's a single-column restriction, in which case `order` was the default cql ordering and this doesn't change anything.
2. it's a multi-column restriction, in which case no index will be eligible and we won't reach this point.

So no bug is introduced. I'll drop the last parameter to make sure there is no unintended change.

Ultimately I think replace_column_def() is miguided. The caller should build a new restriction from scratch, not attempt to do a deep search and replace, so that it is sure about what it is building.

Reply:
Updated with the change removed, it is now compatible with the previous code. We'll fix (or not) independently of this patch.